PushNotifications.md

```diff
@@ -253,19 +253,34 @@ To create local notifications with more content or user interaction options once
 
-#### Foreground notifications
+The SDK does this by implementing [`userNotificationCenter(_:willPresent:withCompletionHandler:)`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649518-usernotificationcenter). This feature is not available on Android.
 
-On iOS, ably-flutter allows you to configure notification sent directly from Ably to be shown when the app is in foreground. To configure if a notification is shown in the foreground:
+#### Data Message / Background Notification
+
+- On iOS, if a notification is also present, it will be shown before `onMessage` or the callback passed to `ably.Push.notificationEvents.setOnBackgroundMessage` is called.
+- On Android, if a notification is also present, it will be shown after these methods finish (or its Future completes).
+- When the app is in the foreground, you can listen to messages using:
 
 ```dart
-ably.Push.notificationEvents.setOnShowNotificationInForeground((message) async {
-  // TODO add logic to show notification based on message contents. 
-  print('Opting to show the notification when the app is in the foreground.');
-  return true;
+ably.Push.notificationEvents.onMessage.listen((remoteMessage) {
+  print('Message was delivered to app while the app was in the foreground: '
+      '$remoteMessage');
 });
 ```
+- This method can be synchronous or `async`.
 
-The SDK does this by implementing [`userNotificationCenter(_:willPresent:withCompletionHandler:)`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649518-usernotificationcenter). This feature is not available on Android.
+- When the app is terminated or in the background, you can listen to messages using:
+
+```dart
+Future<void> _backgroundMessageHandler(ably.RemoteMessage message) async {
+  print('Just received a background message, with:');
+  print('RemoteMessage.Notification: ${message.notification}');
+  print('RemoteMessage.Data: ${message.data}');
+}
+ably.Push.notificationEvents.setOnBackgroundMessage(_backgroundMessageHandler);
+```
+- This method can be synchronous or `async`.
+- On Android, when a message is received whilst the app is terminated, Ably Flutter will launch your application to process the message. Ably Flutter listens to the message by registering a broadcast receiver. This means your `MainActivity` will not be launched, and so your native setup code (e.g. creating method channels) you may have written in your `MainActivity`'s `configureFlutterEngine` method will not be run. If you want your `configureFlutterEngine` even when Ably launches your application, refactor this code into a Flutter package plugin, implementing [`FlutterPlugin`](https://api.flutter.dev/javadoc/io/flutter/embedding/engine/plugins/FlutterPlugin.html) and move your logic inside `onAttachedToEngine`.
```

Comment:
Do we expect library users having an activity `MainActivity`? If so that's good. Otherwise I think it would be better to use a more generic term such as 'the activity you use to connect to Flutter...'

Reply:
"If you want your `configureFlutterEngine` " -> "If you want your `configureFlutterEngine` to run"

Reply:
Thanks: 6323aad8e666f107e65d104d2b9793dbdc893813

Reply:
I'm assuming @ben-xD has decided not to action anything against @ikbalkaya's [opening comment](https://github.com/ably/ably-flutter/pull/227#discussion_r752406314) on this conversation, while I see he has addressed the [second comment](https://github.com/ably/ably-flutter/pull/227#discussion_r752407746).

Therefore, I'm resolving this conversation on @ikbalkaya's behalf.